presto-main/src/main/java/io/prestosql/operator/scalar/FormatFunction.java

```diff
@@ -0,0 +1,227 @@
+import static io.prestosql.spi.type.TinyintType.TINYINT;
+import static io.prestosql.spi.type.TypeSignature.parseTypeSignature;
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static io.prestosql.spi.type.Varchars.isVarcharType;
+import static io.prestosql.type.UnknownType.UNKNOWN;
+import static io.prestosql.util.Reflection.methodHandle;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+
+public final class FormatFunction
+        extends SqlScalarFunction
+{
+    public static final FormatFunction FORMAT_FUNCTION = new FormatFunction();
+    private static final MethodHandle METHOD_HANDLE = methodHandle(FormatFunction.class, "sqlFormat", List.class, ConnectorSession.class, Slice.class, Block.class);
+
+    private FormatFunction()
+    {
+        super(Signature.builder()
+                .kind(SCALAR)
+                .name("format")
+                .typeVariableConstraints(withVariadicBound("T", "row"))
+                .argumentTypes(VARCHAR.getTypeSignature(), parseTypeSignature("T"))
+                .returnType(VARCHAR.getTypeSignature())
+                .build());
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type rowType = boundVariables.getTypeVariable("T");
+
+        List<Function<Block, Object>> converters = mapWithIndex(
+                rowType.getTypeParameters().stream(),
+                (type, index) -> converter(functionRegistry, type, toIntExact(index)))
+                .collect(toImmutableList());
+
+        return new ScalarFunctionImplementation(
+                false,
+                ImmutableList.of(
+                        valueTypeArgumentProperty(RETURN_NULL_ON_NULL),
+                        valueTypeArgumentProperty(RETURN_NULL_ON_NULL)),
+                METHOD_HANDLE.bindTo(converters),
+                true);
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "formats the input arguments using a format string";
+    }
+
+    @UsedByGeneratedCode
+    public static Slice sqlFormat(List<Function<Block, Object>> converters, ConnectorSession session, Slice slice, Block row)
+    {
+        Object[] args = new Object[converters.size()];
+        for (int i = 0; i < args.length; i++) {
+            args[i] = converters.get(i).apply(row);
+        }
+        String format = slice.toStringUtf8();
+
+        try {
+            return utf8Slice(format(session.getLocale(), format, args));
+        }
+        catch (IllegalFormatException e) {
+            String message = e.toString().replaceFirst("^java\\.util\\.(\\w+)Exception", "$1");
```

Comment:
This doesn't work for subclasses like `MissingFormatArgumentException`. Take a look at the test cases for examples of error messages. Or maybe I misunderstand your suggestion?

Reply:
I am concerned that replacing `(\\w+)Exception` with `$1` may produce something that is not self-explanatory, like "Runtime: xxx" (although RTE is java.lang, treat it as an example only).
Instead of assuming (broadly) that every java.util. Exception has a meaningful name (meaningful even when "Exception" is removed), i was trying to narrow down the assumption.
Alternatively, why not keep `"{simple_class_name}: {message}"` for all  java.util. (and java.lang.) exceptions?

Reply:
This is only handling subclasses of `IllegalFormatException` (due to the catch) and only those in `java.util` due to the regex (which means they are limited to a known set since anything in that package is covered by the Java specification).